runtime/memprof.c

```diff
@@ -268,442 +270,219 @@ struct tracked {
   t->unmarshalled = unmarshalled;
   t->promoted = 0;
   t->deallocated = 0;
-  t->in_list = 1;
-  t->in_do_callback = 0;
-  t->next_callback_queue = NULL;
-
-  tracked_set_user_data(t, user_data);
-
+  t->cb_alloc = t->cb_promote = t->cb_dealloc = 0;
+  t->delete = 0;
   return t;
 }
 
-static void try_free_tracked(struct tracked *t)
-{
-  if(!t->in_list && !t->in_do_callback)
-    caml_stat_free(t);
-}
-
-/* Forward declaration for the sole purpose of the assertion in
-   [remove_tracked_and_free]. */
-static struct tracked *last_callback_queue;
-
-/* Remove [t] from the tracked lists if it is still there. Assumes it
-   is not in the callback queue. */
-static void remove_tracked_and_free(struct tracked *t)
+static void mark_deleted(struct tracked* t)
 {
-  CAMLassert(t->next_callback_queue == NULL && t != last_callback_queue);
-  if(t->in_list) {
-    if(t->next != NULL) t->next->prev = t->prev;
-    if(t->prev != NULL) t->prev->next = t->next;
-    else CAMLassert(t == first_tracked_old || t == first_tracked_young);
-    if(t == first_tracked_young) {
-      CAMLassert(t->prev == NULL);
-      first_tracked_young = t->next;
-    }
-    if(t == first_tracked_old)  {
-      CAMLassert(t->prev == NULL);
-      first_tracked_old = t->next;
-    }
-
-    if(t->next_young_user_data != NULL)
-      t->next_young_user_data->prev_young_user_data = t->prev_young_user_data;
-    if(t->prev_young_user_data != NULL) {
-      CAMLassert(t != first_young_user_data);
-      t->prev_young_user_data->next_young_user_data = t->next_young_user_data;
-    } else if(t == first_young_user_data) {
-      first_young_user_data = t->next_young_user_data;
-    }
-
-    t->in_list = 0;
-  }
-  try_free_tracked(t);
+  t->delete = 1;
+  t->user_data = Val_unit;
+  t->block = Val_unit;
 }
 
-/**** Handling callbacks. ****/
-/* When an event occurs (allocation, promotion or deallocation), we
-   cannot immediately call the callback, because we may not be in a
-   context where the execution of arbitrary OCaml code is
-   allowed. These functions make it possible to register a callback in
-   a todo-list so that the call is performed when possible. */
-
-/* The TODO list is represented as a singly linked list. All its
-   elements are members of the [tracked_old]/[tracked_young] list. */
-static struct tracked *first_callback_queue = NULL,
-  *last_callback_queue = NULL;
-
-/* Register a callback for the given tracked block. If another
-   callback for this block is pending or running in another thread,
-   then this function does not do anything. In this case, we let
-   [do_callback] register this new callback.
-
-   This function does not call the GC. This is important since it is
-   called when allocating a block using [caml_alloc_shr]: The new
-   block is allocated, but not yet initialized, so that the heap
-   invariants are broken. */
-static void register_postponed_callback(struct tracked *t, enum cb_kind kind)
+static void check_exn(value res)
```

Comment:
Minor nitpick: so far, all C functions ending in `_exn` return a value that must be tested with `Is_exception_result`. It is quite convenient to know this from the name of the function alone. I know that no such convention has been discussed, but do you like the idea? I suggest the name `check_exception` instead, similarly to `caml_raise_if_exception` (which I avoided naming `caml_raise_if_exn`). (I know this is a very minor detail for a local function.)

Reply:
I like the idea. I still need to do some work on this peice of code in order to make it reentrant and possibly make it exception-safe (these seems to be two faces of the same problem). Don't worry, this will be fixed.